### PR TITLE
stat/distuv: add Score method to Triangle and Uniform

### PR DESCRIPTION
**Pull request by romanwerpachowski**
Also tests and fix Laplace Score.

Please take a look.

### CODE REVIEW

**Comment by kortschak** [comment]
For the sake of diff reduction, making the name changes in place in this PR would be best. The order of the functions to match alphabetical ordering can happen later.

The failures are due to the now-unexported methods not being tested. Those would need to be added.

**Comment by romanwerpachowski** [comment]
PTAL.

File: stat/distuv/triangle.go [hunk 1]
```diff
@@ -79,6 +79,7 @@ func (t Triangle) Median() float64 {
 	if t.c >= (t.a+t.b)/2 {
 		return t.a + math.Sqrt((t.b-t.a)*(t.c-t.a)/2)
 	}
+
```
Comment (kortschak) [inline stat/distuv/triangle.go:82]: Please revert this blank line insertion.

In general the convention for returning if-else constructs is `if cond { return res }; return altres`, which in other languages would be written as `if cond { return res } else { return altres }` (there are historical reasons for this relating to return value compiler checks that are not relevant in modern Go code, but the convention remains). This means that the alternative branch is part of the stanza, just as is would be if there were an explicit `else`.

File: stat/distuv/triangle.go [hunk 2]
```diff
@@ -134,6 +135,66 @@ func (t Triangle) Rand() float64 {
...
+// of the log-likelihood at x with respect to the parameters
+//  (∂/∂θ) log(p(x;θ))
+// If deriv is non-nil, len(deriv) must equal the number of parameters otherwise
+// Score will panic, and the derivative is stored in-place into deriv. If deriv
+// is nil a new slice will be allocated and returned.
+//
+// The order is [∂LogProb / ∂Mu, ∂LogProb / ∂Sigma].
+//
+// For more information, see https://en.wikipedia.org/wiki/Score_%28statistics%29.
+func (t Triangle) Score(deriv []float64, x float64) []float64 {
```
Comment (kortschak) [inline stat/distuv/triangle.go:149]: This is a new addition; it and its tests should be another PR (possibly the fix to `Laplace` too).

File: stat/distuv/uniform.go [hunk 3]
```diff
@@ -116,6 +120,40 @@ func (u Uniform) Rand() float64 {
...
+// of the log-likelihood at x with respect to the parameters
+//  (∂/∂θ) log(p(x;θ))
+// If deriv is non-nil, len(deriv) must equal the number of parameters otherwise
+// Score will panic, and the derivative is stored in-place into deriv. If deriv
+// is nil a new slice will be allocated and returned.
+//
+// The order is [∂LogProb / ∂Mu, ∂LogProb / ∂Sigma].
+//
+// For more information, see https://en.wikipedia.org/wiki/Score_%28statistics%29.
+func (u Uniform) Score(deriv []float64, x float64) []float64 {
```
Comment (kortschak) [inline stat/distuv/uniform.go:134]: Same here.

File: stat/distuv/laplace.go [hunk 4]
```diff
@@ -203,7 +203,7 @@ func (l Laplace) Score(deriv []float64, x float64) []float64 {
 		deriv[0] = math.NaN()
 	}
 
-	deriv[1] = math.Abs(diff)/(l.Scale*l.Scale) - 0.5/(l.Scale)
+	deriv[1] = math.Abs(diff)/(l.Scale*l.Scale) - 1/l.Scale
```
Comment (kortschak) [inline stat/distuv/laplace.go:206]: I would do this in the PR that adds `Score` to the other distributions.

File: stat/distuv/triangle.go (hunk 2 above)
Comment (romanwerpachowski) [reply stat/distuv/triangle.go:149]: If I don't add the Score methods, I can't test parameters and setParameters using the existing testDerivativeParam function.

File: stat/distuv/laplace.go (hunk 4 above)
Comment (romanwerpachowski) [reply stat/distuv/laplace.go:206]: If I don't add this fix, the function which tests parameters and setParameters fails.

File: stat/distuv/triangle.go (hunk 2 above)
Comment (kortschak) [reply stat/distuv/triangle.go:149]: Mmm, OK. I think that changes the focus of the PR then, which is fine. It would become

```
stat/distuv: add Score method to Triangle and Uniform

Also tests and fix Laplace Score.
```

Comment (romanwerpachowski) [reply stat/distuv/triangle.go:149]: Done.

File: stat/distuv/triangle.go (hunk 1 above)
Comment (romanwerpachowski) [reply stat/distuv/triangle.go:82]: Done.

**Comment by romanwerpachowski** [comment]
PTAL.

File: stat/distuv/laplace.go [hunk 5]
```diff
@@ -199,7 +203,7 @@ func (l Laplace) Score(deriv []float64, x float64) []float64 {
...
```
[the remark this answers is not in the document]
Comment (romanwerpachowski) [reply stat/distuv/laplace.go:206]: Good catch, and you actually helped me to find another bug ;-)

[the remark this answers is not in the document]
Comment (romanwerpachowski) [reply stat/distuv/laplace.go:204]: PTAL.

[the remark this answers is not in the document]
Comment (kortschak) [reply stat/distuv/laplace.go:204]: Yeah, I was thinking that might be the case.

File: stat/distuv/laplace.go [hunk 6]
```diff
@@ -192,14 +196,12 @@ func (l Laplace) Score(deriv []float64, x float64) []float64 {
 		deriv[0] = 1 / l.Scale
 	} else if diff < 0 {
 		deriv[0] = -1 / l.Scale
-	} else if diff == 0 {
-		deriv[0] = 0
 	} else {
 		// must be NaN
 		deriv[0] = math.NaN()
 	}
```
Comment (kortschak) [inline stat/distuv/laplace.go:202]: Is this described as the convention somewhere?

Comment (romanwerpachowski) [reply stat/distuv/laplace.go:202]: I didn't find anything like that. If the derivative doesn't exist, we should return NaN IMHO.
